File: dotnet/Lexy/Lexy.Compiler/Parser/SourceReference.cs
using System;

namespace Lexy.Compiler.Parser
{
    public class SourceReference
    {
        private readonly int? lineNumber;
        private readonly int? characterNumber;

        public SourceFile File { get; }

        public SourceReference(SourceFile file, int? lineNumber, int? characterNumber)
        {
            File = file ?? throw new ArgumentNullException(nameof(file));
            this.characterNumber = characterNumber;
            this.lineNumber = lineNumber;
        }

        public override string ToString() => $"{File.FileName}({lineNumber}, {characterNumber})";
    }
}

File: typescipt/src/parser/sourceReference.ts


export class SourceReference {
   private readonly number? characterNumber;
   private readonly number? lineNumber;

   public SourceFile File

   constructor(file: SourceFile, lineNumber: int?, characterNumber: int?) {
     File = file ?? throw new Error(nameof(file));
     this.characterNumber = characterNumber;
     this.lineNumber = lineNumber;
   }

   public override toString(): string {
     return $`{File.FileName}({lineNumber}, {characterNumber})`;
   }
}
